pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+	config.SetDefaults()
+
+	if err := config.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid attacher configuration: %w", err)
+	}
+
+	ua := &UprobeAttacher{
+		name:                 name,
+		config:               config,
+		fileRegistry:         utils.NewFileRegistry(name),
+		manager:              mgr,
+		onAttachCallback:     onAttachCallback,
+		pathToAttachedProbes: make(map[string][]manager.ProbeIdentificationPair),
+		done:                 make(chan struct{}),
+		inspector:            inspector,
+	}
+
+	utils.AddAttacher(name, ua)
+
+	return ua, nil
+}
+
+// SetRegistry allows changing the file registry used by the attacher. This is useful for testing purposes, to
+// replace the registry with a mock object
+func (ua *UprobeAttacher) SetRegistry(registry FileRegistry) {
+	ua.fileRegistry = registry
+}
+
+// handlesLibraries returns whether the attacher has rules configured to attach to shared libraries.
+// It caches the result to avoid recalculating it every time we are attaching to a PID.
+func (ua *UprobeAttacher) handlesLibraries() bool {
+	if ua.handlesLibrariesCached != nil {
+		return *ua.handlesLibrariesCached
+	}
+
+	result := false
+	for _, rule := range ua.config.Rules {
+		if rule.canTarget(AttachToSharedLibraries) {
+			result = true
+			break
+		}
+	}
+	ua.handlesLibrariesCached = &result
+	return result
+}
+
+// handlesLibraries returns whether the attacher has rules configured to attach to executables directly
+// It caches the result to avoid recalculating it every time we are attaching to a PID.
+func (ua *UprobeAttacher) handlesExecutables() bool {
+	if ua.handlesExecutablesCached != nil {
+		return *ua.handlesExecutablesCached
+	}
+
+	result := false
+	for _, rule := range ua.config.Rules {
+		if rule.canTarget(AttachToExecutable) {
+			result = true
+			break
+		}
+	}
+	ua.handlesExecutablesCached = &result
+	return result
+}
+
+// Start starts the attacher, attaching to the processes and libraries as needed
+func (ua *UprobeAttacher) Start() error {
+	var cleanupExec, cleanupExit func()
+	procMonitor := monitor.GetProcessMonitor()
+	err := procMonitor.Initialize(ua.config.ProcessMonitorEventStream)
+	if err != nil {
+		return fmt.Errorf("error initializing process monitor: %w", err)
+	}
+
+	if ua.handlesExecutables() {
+		cleanupExec = procMonitor.SubscribeExec(ua.handleProcessStart)
+	}
+	// We always want to track process deletions, to avoid memory leaks
+	cleanupExit = procMonitor.SubscribeExit(ua.handleProcessExit)
+
+	if ua.config.PerformInitialScan {
+		// Initial scan only looks at existing processes, and as it's the first scan
+		// we don't have to track deletions
+		err := ua.Sync(true, false)
+		if err != nil {
+			return fmt.Errorf("error during initial scan: %w", err)
+		}
+	}
+
+	if ua.handlesLibraries() {
+		if !sharedlibraries.IsSupported(ua.config.EbpfConfig) {
+			return errors.New("shared libraries tracing not supported for this platform")
+		}
```

Comment:
nit: I do recommend to re-order the code

We can run `ua.Sync` and then to abort, so I do recommend to put the heavy operations as late as possible to prevent "redundant" processing

Reply:
Changed, put Sync as the last operation before the main loop, let me know if that's what you meant.